### ISSUE TITLE
Type instabilities all over the place

### ISSUE THREAD

**Issue by dpsanders** [dpsanders/ValidatedNumerics.jl#189]
Simple example on master (50eb702310545f5):

```
julia> a = 0..1
[0, 1]

julia> @code_warntype a^2
Variables:
  #self#::Base.#^
  a::ValidatedNumerics.Interval{Float64}
  x::Int64

Body:
  begin
      return (ValidatedNumerics.convert)(ValidatedNumerics.Interval{Float64},$(Expr(:invoke, LambdaInfo for ^(::ValidatedNumerics.Interval{BigFloat}, ::Int64), :(ValidatedNumerics.^), :($(Expr(:invoke, LambdaInfo for big53(::ValidatedNumerics.Interval{Float64}), :(ValidatedNumerics.big53), :(a)))), :(x))))::Any
  end::Any
```

Also `@code_warntype sqrt(a)`.

Much of this would be fixed by #188, e.g. `sqrt(a)` is then type stable.

**Comment by lbenet** [comment]
Is this related to #91?

**Comment by dpsanders** [comment]
No, #91 is type stable on Julia 0.5 and 0.6. 

I am not sure exactly where the problem is coming from, but I think it is to do with mis-using macros by passing in the type parameter explicitly in `@round(T, ...)`. This is what #188 solves, by letting Julia use dispatch instead.

**Comment by lbenet** [comment]
You are right, #91 has nothing to do here...

Do you have other examples about the type unstability?

I agree with you that it may be related with the use and misuse of macros. Considering your example, `@code_warntype ValidatedNumerics.big53(a)` is type stable, but `@code_warntype (ValidatedNumerics.big53(a))^2` isn't.

I'll try to take a look on #188 later. Another possibility could be to add type declarations. What do you think about this?

**Comment by dpsanders** [comment]
I think #188 is progressing well, so I don't think that type declarations are required, or maybe only in certain key places.

**Comment by dpsanders** [comment]
Fixed by #188 .
